django/thunderstore/repository/migrations/0031_add_namespaces_to_existing_packages.py

```diff
@@ -0,0 +1,31 @@
+# Generated by Django 3.1.7 on 2021-11-04 20:32
+
+from django.db import migrations
+
+
+def forwards(apps, schema_editor):
+    Package = apps.get_model("repository", "Package")
+    packages = Package.objects.all()
+    for p in packages:
```

Comment:
I'm fairly confident you could make this whole migration far more efficient by leveraging annotations and/or subqueries:
- https://docs.djangoproject.com/en/dev/ref/models/expressions/#subquery-expressions
- https://www.paulox.net/2018/10/01/updating-a-django-queryset-with-annotation-and-subquery/

That being said, it might not be worth the time investment here, but feel free to use it as a learning experience if you want.

Also I would still encourage checking that the team and namespace names match here, e.g. by using `namespaces.filter(name__iexact=p.owner.name).first()` instead of just getting the first one. Also it would be a good idea to fail if no namespace is found for a reason or another, as Package.namespace _should_ be a non-nullable field (and will need to be migrated to be such a field in the future)

Reply:
Ok, so I tried introducing Subqueries and OuterRefs into this, but it just gets a bit too complicated for the use case. I did tho learn a bit on, how to use Subqueries and OuterRefs to minimize the amount of queries that are used.

On this specific use case, I ended up in the same situation with the amount of queries as in the beginning, since I had to first query all the Packages with matching Namespace object existing to their owners name. And then the rest that did not have em. Perhaps I could have put this value of the Namespace existing or not, to some annotated variable and then updating all of the packages at once; by either using the existing one as the namespace or by creating a new one if needed later. (But as I stated before, it just gets a bit too complicated in the code IMO)

All in all, I now know how to apply this to some part of code, that doesn't have to create new objects and make it better, faster, stronger.

Reply:
Sounds good enough to me 👍 It's not as relevant in a migration (as those are one-off operations), but it's worth getting familiar with the feature to know where it can be leveraged in other scenarios.